includes/discovery/ipv6-addresses.inc.php

```diff
@@ -27,7 +27,7 @@
         }
     }
 
-    if (empty($oids)) {
+    if (empty($oids) || empty($valid)) {
```

Comment:
Is this the right check to make as `$valid` will only be relevant for the last discover_process_ipv6 call in the foreach loop.

What's the output of `$oids` when we believe it's not valid?

Reply:
`$valid` is actually an array that gets passed by reference to, and incrementally populated by, `discover_process_ipv6()`

https://github.com/librenms/librenms/blob/389cdcbcb22be03559f0980ef41d1a3a24a3429c/includes/discovery/functions.inc.php#L654

https://github.com/librenms/librenms/blob/389cdcbcb22be03559f0980ef41d1a3a24a3429c/includes/discovery/functions.inc.php#L705-L707

In the Juniper case - `$oids` is populated due to `IP-MIB::ipAddressIfIndex.ipv6` being populated, but `discover_process_ipv6()` fails due to `IP-MIB::ipAddressPrefix.ipv6` being garbage.

```php
    }

    var_dump(["oids" => $oids, "valid" => $valid]); // added for debugging

    if (empty($oids) || empty($valid)) {
```

```
array(2) {
  ["oids"]=>
  array(1) {
    ["ipv6"]=>
    array(3) {
      ["20:01:0d:b8:de:ad:be:ef:ca:fe:00:00:00:00:00:01"]=>
      array(3) {
        ["IP-MIB::ipAddressIfIndex"]=>
        string(3) "505"
        ["IP-MIB::ipAddressOrigin"]=>
        string(6) "manual"
        ["IP-MIB::ipAddressPrefix"]=>
        string(23) "SNMPv2-SMI::zeroDotZero"
      }
      ["20:01:de:ad:be:ef:ca:fe:00:00:00:00:00:00:00:10"]=>
      array(3) {
        ["IP-MIB::ipAddressIfIndex"]=>
        string(3) "505"
        ["IP-MIB::ipAddressOrigin"]=>
        string(6) "manual"
        ["IP-MIB::ipAddressPrefix"]=>
        string(23) "SNMPv2-SMI::zeroDotZero"
      }
      ["fe:80:00:00:00:00:00:00:7a:19:f7:ff:fe:9e:81:83"]=>
      array(3) {
        ["IP-MIB::ipAddressIfIndex"]=>
        string(3) "505"
        ["IP-MIB::ipAddressOrigin"]=>
        string(9) "linklayer"
        ["IP-MIB::ipAddressPrefix"]=>
        string(23) "SNMPv2-SMI::zeroDotZero"
      }
    }
  }
  ["valid"]=>
  array(0) {
  }
}
```

Reply:
for comparison, valid from IOS:
```
array(2) {
  ["oids"]=>
  array(1) {
    ["ipv6"]=>
    array(2) {
      ["20:01:0d:b8:de:ad:be:ef:ca:fe:00:00:00:00:00:20"]=>
      array(3) {
        ["IP-MIB::ipAddressIfIndex"]=>
        string(1) "1"
        ["IP-MIB::ipAddressOrigin"]=>
        string(6) "manual"
        ["IP-MIB::ipAddressPrefix"]=>
        string(93) "IP-MIB::ipAddressPrefixOrigin[1][ipv6]["20:01:0d:b8:de:ad:be:ef:00:00:00:00:00:00:00:00"][64]"
      }
      ["20:01:de:ad:be:ef:ca:fe:00:00:00:00:00:00:00:20"]=>
      array(3) {
        ["IP-MIB::ipAddressIfIndex"]=>
        string(1) "2"
        ["IP-MIB::ipAddressOrigin"]=>
        string(6) "manual"
        ["IP-MIB::ipAddressPrefix"]=>
        string(93) "IP-MIB::ipAddressPrefixOrigin[2][ipv6]["20:01:de:ad:be:ef:ca:fe:00:00:00:00:00:00:00:00"][64]"
      }
    }
  }
  ["valid"]=>
  array(1) {
    ["ipv6"]=>
    array(2) {
      ["2001:0db8:dead:beef:cafe:0000:0000:0020/64-86"]=>
      int(1)
      ["2001:dead:beef:cafe:0000:0000:0000:0020/64-87"]=>
      int(1)
    }
  }
}
```